packages/astro/src/actions/runtime/middleware.ts

```diff
@@ -47,15 +46,15 @@ export const onRequest = defineMiddleware(async (context, next) => {
 		},
 	};
 	Object.defineProperty(locals, '_actionsInternal', { writable: false, value: actionsInternal });
-	return next();
+	return ApiContextStorage.run(context, () => next());
```

Comment:
Just a word of warning... when users use the rewrites APIs inside the middleware, under the hood, we create a new `APIContext`; there's a chance that this `context` could be outdated in case users decide to use the rewrite APIs:

```js
next("/new-url") // new API, it will generate a new context with a new `Request`, and the URL is `/new-url`
```

Reply:
Ah, that's a good catch! I _think_ applying the `APIContext` to this case would resolve our issue, since it only gets called when a rewrite is detected: https://github.com/withastro/astro/pull/11088/files#diff-a07f25105bdef494ce45893f063230e218d44b36cd7db7c9c8d6447cb03db136R19

I'll need a unit test to confirm

Reply:
It's also true that having a rewrite inside an action feels weird, not sure there's a valid case for that. I could create a stackblitz after the release and see how that works

Reply:
Added a test. It works!